File: leetcode/C++/count-complete-tree-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getHeight(TreeNode* node){
        int l=-1;
        for(; node!=nullptr; node=node->left, ++l) {}
        return l;
    }
    int countNodes(TreeNode* root) {
        if(root==nullptr) return 0;
        int sol=0,h=0;
        for(h=getHeight(root), sol=(1<<h); h>0; --h){
            if(getHeight(root->right)+1==h){
                root=root->right;
                sol+=(1<<(h-1));
            }
            else root=root->left;
        }
        return sol;
    }
};

/*
생각보다 o(N) 보다 작은 풀이가 떠오르지가 않음
*/

File: leetcode/Python/count-complete-tree-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        def getHeight(node):
            l = -1
            while node:
                node, l = node.left, l + 1
            return l
        h, sol = 0, 0
        if (h:=getHeight(root)) == -1:
            return 0
        sol = 1<<h
        for i in range(h, 0, -1):
            if getHeight(root.right)+1 == i:
                sol, root = sol + (1<<i-1), root.right
            else:
                root = root.left
        return sol